<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>ASAFE (Ancestry Specific Allele Frequency Estimation) by BiostatQian</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>ASAFE (Ancestry Specific Allele Frequency Estimation)</h1>
        <h2>Estimating the frequency of an allele for a bi-allelic marker in a 3-way admixed population:
        * Currently being updated [June 4, 2016]* </h2>
        <a href="https://github.com/BiostatQian/ASAFE" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h2>
<a id="what-is-asafe" class="anchor" href="#what-is-asafe" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>What is ASAFE?</h2>

<p>The ASAFE EM algorithm provides maximum likelihood estimates of ancestry-specific allele frequencies at a bi-allelic marker, given local ancestries and genotypes at the marker. It deals with uncertainty in the phase (i.e. order) of a local ancestry pair relative to a genotype. </p>

<p>ASAFE was motivated by genome-wide association studies of Hispanics performed by the University of Washington Genetic Analysis Center as part of the Hispanic Community Health Study (<a href="https://www2.cscc.unc.edu/hchs/">https://www2.cscc.unc.edu/hchs/</a>), and applies to any 3-way admixed diploid population, not just humans.</p>

<h2>
<a id="picture-showing-how-asafe-fits-into-a-genetic-analysis-workflow" class="anchor" href="#picture-showing-how-asafe-fits-into-a-genetic-analysis-workflow" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a> How does ASAFE fit into a genetic analysis workflow? Show me the picture. </h2>

<p>In the top folder of the ASAFE R package is a file, ASAFE_Visual.pdf, illustrating how ASAFE fits into a genetic analysis workflow: </p>

<p><a href="https://github.com/BiostatQian/ASAFE/raw/master/inst/ASAFE_Visual.pdf">Picture of a Workflow using ASAFE</a></p>

<p>Sharon Browning has some code that does the steps in the diagram involving phasing of genotypes and running of RFMIX.
It is the part of the following script above the line that says "Apply masking": <a href="http://faculty.washington.edu/sguy/local_ancestry_pipeline/rfmix_mds_pipeline">http://faculty.washington.edu/sguy/local_ancestry_pipeline/rfmix_mds_pipeline</a>.
I've extracted the relevant part of the script here: </p>

<p><a href="https://github.com/BiostatQian/ASAFE/raw/master/inst/scripts/bgl_then_rfmix.sh">BEAGLE then RFMIX</a></p>

You'd then write your own script to go from text file output from RFMIX to text file input to ASAFE.

<h2>
<a id="how-to-use-the-asafe-r-package" class="anchor" href="#how-to-use-the-asafe-r-package" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>How do I use the ASAFE R Package?</h2>

<p>In the ASAFE R package folder, see inst/doc/ASAFE.pdf for instructions on how to use the package. Cheers!</p>

<h2>
<a id="paper-reference" class="anchor" href="#paper-reference" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Paper Reference
</h2>

<p>Zhang QS, Browning BL, and Browning SR (2016) “Ancestry Specific Allele Frequency Estimation.” Bioinformatics.</p>

<p> There's a typo in the supplement. Tables are called "Table 1", "Table 2", "Supplementary Table 1", and "Supplementary Table 2", instead of Tables 1, 2, 3, and 4. I've changed comments in the code to hopefully make the distinction clear. </p>

<h2>
<a id="paper-reference" class="anchor" href="#paper-reference" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Changes and Feedback
</h2>

<p>
[Changes made June 4, 2016]
</p>

<p>
(1) Removed function em() from vignette, because a user would likely only be interested in using function algorithm_1snp(), which calls em().
</p>

<p>
(2) Change variable names in estep.R, mstep.R, and em.R to match the supplement.
</p>

<h2>
<a id="contributors" class="anchor" href="#contributors" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Contributors
</h2>


<p> If your question is not answered on this page or in the vignette, feel free to contact me with package-related questions. </p>

<p>Qian Sophia Zhang (<a href="mailto:qszhang@uw.edu">qszhang@uw.edu</a>)</p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/BiostatQian/ASAFE/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/BiostatQian/ASAFE/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/BiostatQian/ASAFE"></a> is maintained by <a href="https://github.com/BiostatQian">BiostatQian</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
